example/ipsec_api/odp_ipsec.c

```diff
@@ -0,0 +1,1248 @@
+ */
+static
+pkt_disposition_e do_input_verify(odp_packet_t pkt,
+				  pkt_ctx_t *ctx EXAMPLE_UNUSED)
+{
+	if (odp_unlikely(odp_packet_has_error(pkt)))
+		return PKT_DROP;
+
+	if (!odp_packet_has_eth(pkt))
+		return PKT_DROP;
+
+	if (!odp_packet_has_ipv4(pkt))
+		return PKT_DROP;
+
+	return PKT_CONTINUE;
+}
+
+/**
+ * Packet Processing - Route lookup in forwarding database
+ *
+ * @param pkt  Packet to route
+ * @param ctx  Packet process context
+ *
+ * @return PKT_CONTINUE if route found else PKT_DROP
+ */
+static
+pkt_disposition_e do_route_fwd_db(odp_packet_t pkt, pkt_ctx_t *ctx)
+{
+	odph_ipv4hdr_t *ip = (odph_ipv4hdr_t *)odp_packet_l3_ptr(pkt, NULL);
+	fwd_db_entry_t *entry;
+
+	entry = find_fwd_db_entry(odp_be_to_cpu_32(ip->dst_addr));
+
+	if (entry) {
+		uint32_t l3_offset = odp_packet_l3_offset(pkt);
+
+		if (l3_offset > sizeof(odph_ethhdr_t))
+			odp_packet_pull_head(pkt,
+					     l3_offset - sizeof(odph_ethhdr_t));
+		else
+			odp_packet_push_head(pkt,
+					     sizeof(odph_ethhdr_t) - l3_offset);
+
+		memcpy(&ctx->eth.dst, entry->dst_mac, ODPH_ETHADDR_LEN);
+		memcpy(&ctx->eth.src, entry->src_mac, ODPH_ETHADDR_LEN);
+		ctx->eth.type = odp_cpu_to_be_16(ODPH_ETHTYPE_IPV4);
+
+		if (args->appl.mode != ODP_IPSEC_OP_MODE_INLINE) {
+			odp_packet_l2_offset_set(pkt, 0);
+			odp_packet_copy_from_mem(pkt, 0, ODPH_ETHHDR_LEN,
+						 &ctx->eth);
+		}
+
+		ctx->pktio = entry->pktio;
+		ctx->pktout = entry->pktout;
+
+		return PKT_CONTINUE;
+	}
+
+	return PKT_DROP;
+}
+
+/**
+ * Packet Processing - Input IPsec packet classification
+ *
+ * Verify the received packet has IPsec headers and a match
+ * in the IPsec cache, if so issue crypto request else skip
+ * input crypto.
+ *
+ * @param pkt   Packet to classify
+ * @param ctx   Packet process context
+ *
+ * @return PKT_CONTINUE if done else PKT_POSTED
+ */
+static
+pkt_disposition_e do_ipsec_in_classify(odp_packet_t *ppkt)
+{
+	odph_ipv4hdr_t *ip = (odph_ipv4hdr_t *)odp_packet_l3_ptr(*ppkt, NULL);
+	odph_ahhdr_t *ah = NULL;
+	odph_esphdr_t *esp = NULL;
+	ipsec_cache_entry_t *entry;
+	odp_ipsec_in_param_t in_param;
+	int rc;
+
+	/* Check IP header for IPSec protocols and look it up */
+	locate_ipsec_headers(ip, &ah, &esp);
+	if (!ah && !esp)
+		return PKT_CONTINUE;
+	entry = find_ipsec_cache_entry_in(odp_be_to_cpu_32(ip->src_addr),
+					  odp_be_to_cpu_32(ip->dst_addr),
+					  ah,
+					  esp);
+	if (!entry)
+		return PKT_CONTINUE;
+
+	memset(&in_param, 0, sizeof(in_param));
+	if (args->appl.lookup) {
+		in_param.num_sa = 0;
+		in_param.sa = NULL;
+	} else {
+		in_param.num_sa = 1;
+		in_param.sa = &entry->ipsec_sa;
+	}
+
+	/* Issue crypto request */
+	if (args->appl.mode != ODP_IPSEC_OP_MODE_SYNC) {
+		rc = odp_ipsec_in_enq(ppkt, 1, &in_param);
+		if (rc <= 0)
+			return PKT_DROP;
+
+		return PKT_POSTED;
+	} else {
+		int out = 1;
+
+		rc = odp_ipsec_in(ppkt, 1, ppkt, &out, &in_param);
+		if (rc <= 0)
+			return PKT_DROP;
+
+		return PKT_CONTINUE;
+	}
```

Comment:
Checkpatch complains about this construct. Prefers no else after return:
```
WARNING: else is not generally useful after a break or return
#795: FILE: example/ipsec_api/odp_ipsec.c:603:
+		return PKT_POSTED;
+	} else {
```

Reply:
And I prefer current code. It is cleaner from my point of view.
```c
if (!SYNC) {
  odp_ipsec_in_enq();
  return PKT_POSTED:
} else {
  odp_ipsec_in();
  return PKT_CONTINUE;
}
```